File: August/Week1/Day7/Submissions/verticalOrder.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
    const map = new Map();
    let queue = [[root, 0]];
    
    while(queue.length) {
        const cols = new Map(), next = [];
        
        for(let [node, x] of queue) {
            if(!cols.has(x)) cols.set(x, [node.val]);
            else cols.get(x).push(node.val);
            
            if(node.left) next.push([node.left, x-1]);
            if(node.right) next.push([node.right, x+1]);
        }
        
        for(let [x, val] of cols) {
            if(!map.has(x)) map.set(x, []);
            map.get(x).push(...val.sort((a, b) => a - b));
        }
        queue = next;
    }
    return [...[...map.entries()].sort((a, b) => a[0] - b[0]).map(x => x[1])]
};

File: August/Week1/Day7/workArea/verticalOrder.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        dic = defaultdict(list)
        self.min_l, self.max_l = float("inf"), -float("inf")
        def dfs(root, lvl_h, lvl_v):
            self.min_l = min(lvl_h, self.min_l)
            self.max_l = max(lvl_h, self.max_l)
            dic[lvl_h].append((lvl_v, root.val))
            if root.left:  dfs(root.left,  lvl_h-1, lvl_v+1)
            if root.right: dfs(root.right, lvl_h+1, lvl_v+1)
        
        dfs(root, 0, 0)
        res = []
        for i in range(self.min_l, self.max_l + 1):
            res += [[j for i,j in sorted(dic[i])]]
        return res